File: 135-candy/candy.cpp
class Solution {
public:
    int candy(vector<int>& ratings) {
        int n = ratings.size(), candy=n;

        int i=1;
        while(i<n){
            if(ratings[i] == ratings[i-1]){i++;continue;}

            int peak=0;
            while(i<n && ratings[i] > ratings[i-1]){
                peak++;
                candy += peak;
                i++;
            }

            if(i==n)return candy;

            int dip=0;
            while(i<n && ratings[i]<ratings[i-1]){
                dip++;
                candy += dip;
                i++;
            }

            candy -= min(dip,peak);
        }

        return candy;
    }
};

File: 135-candy/candy.java
class Solution {
    public int candy(int[] ratings) {
        int n = ratings.length;

        int[] L2R = new int[n];
        int[] R2L = new int[n];

        Arrays.fill(L2R,1);
        Arrays.fill(R2L,1);

        for(int i=1; i<n; i++){
            if(ratings[i] > ratings[i-1])L2R[i] = L2R[i-1]+1;
        }

        for(int i=n-2; i>=0; i--){
            if(ratings[i] > ratings[i+1])R2L[i] = R2L[i+1]+1;
        }

        int candy=0;
        for(int i=0; i<n; i++)candy += Math.max(L2R[i],R2L[i]);

        return candy;
    }
}